### PR TITLE
Dependency injection identified by name and type

### LINKED ISSUE #18: Allow overwriting default dependency injectors with same return type

### PR DESCRIPTION
**Pull request by Sinitax**
closes #18

### CODE REVIEW

**Comment by ldruschk** [comment]
Is there a convenient way to get two objects from the same injector? This is used for example here to get two client sessions:
https://github.com/ldruschk/bambi-service-stonksexchange/blob/94f4460ef1d8ac8803e5f4faffbf3c0e5dd4b1d6/checker/checker.py#L51

File: enochecker3/enochecker.py [hunk 1]
```diff
@@ -388,6 +390,8 @@ def decorator(f: Callable[..., Any]) -> None:
 
             self._dependency_injections[key] = f
 
+            return f
```
Comment (Sinitax) [inline enochecker3/enochecker.py:393]: just need to return the wrapped function so the next decorator can pick it up again

**Comment by Sinitax** [comment]
^ you can always just inject "injector" and then get your sessions that way:

```
@checker.putflag(0)
async def putflag_test(task: PutflagCheckerTaskMessage, injector: DependencyInjector, db: ChainDB, logger: LoggerAdapter) -> None:
    session_a = injector.get("session", AsyncClient)
    session_b = injector.get("session", AsyncClient)
    await register_user(session_a)
    logger.debug("registered user for session_a")
    (username_b, password_b) = await register_user(session_b)
    logger.debug("registered user for session_b")

    await send_message(session_a, username_b, task.flag)

    await db.set("credentials", (username_b, password_b))
```

if you want a "convenient" way, we could split the name by something like '_' and use only the first part for resolution?

**Comment by ldruschk** [comment]
how about we make the name optional, as long as there is only one dependency injection per type?

that would probably require some rework, but that should also retain backwards compatibility (although that would not really be necessary, I would prefer it)

File: tests/test_dependency_injection.py [hunk 2]
```diff
@@ -21,13 +21,31 @@ def havoc_task() -> HavocCheckerTaskMessage:
...
+    @checker.register_dependency()
+    def inject_string() -> str:
+        return "123"
+
+    @checker.register_dependency("special")
+    def inject_special_string() -> str:
+        return "456"
+
+    @checker.havoc(0)
+    async def havoc(a: str, b: str, special: str):
```
Comment (ldruschk) [inline tests/test_dependency_injection.py:47]: just to clarify, this means we can mix unnamed dependency injections with named dependency injections with the same type and when injecting, if no name matches one of the named dependencies, we use the unnamed dependency injection?

Comment (Sinitax) [reply tests/test_dependency_injection.py:47]: yep, the fallback when the name does not match is to look for a 'generic' injector (empty name)

Comment (Sinitax) [reply tests/test_dependency_injection.py:47]: hmm, the syntax is a bit different so this is actually still not backwards compatible

Comment (Sinitax) [reply tests/test_dependency_injection.py:46]: Tested [9dcc1c7](https://github.com/enowars/enochecker3/pull/27/commits/9dcc1c7692d1c5ca8c47d1ff10efb4fc7378e86f) against enowars5-stldoctor

**Pull request merged**